lib/fast-boot.ts

```diff
@@ -0,0 +1,110 @@
+/**
+ * @license
+ * Copyright 2021 Balena Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * This module sets up the `fast-boot2` module, including testing whether
+ * we have permissions over the cache file before even attempting to load
+ * fast boot.
+ * AVOID LOADING OTHER MODULES HERE AS MUCH AS POSSIBLE, AND KEEP ALL
+ * FUNCTIONS SYNCHRONOUS!
+ */
+
+let fastBootStarted = false;
+
+export function start() {
+	if (fastBootStarted) {
+		return;
+	}
+	const os = require('os') as typeof import('os');
+	const path = require('path') as typeof import('path');
+	const dotBalena = process.platform === 'win32' ? '_balena' : '.balena';
+	// TODO: take into account `~/.balenarc.yml` or `./balenarc.yml`,
+	// without hurting performance at this early loading stage.
+	const dataDir = path.normalize(
+		process.env.BALENARC_DATA_DIRECTORY || path.join(os.homedir(), dotBalena),
+	);
+	// Consider that the CLI may be installed to a folder owned by root
+	// such as `/usr[/local]/lib/balena-cli`, while being executed by
+	// a regular user account.
+	const cacheFile = path.join(dataDir, 'cli-module-cache.json');
```

Comment:
The code comment above, and this PR's description, explains the motivation for changing the location of the module cache file:
* Before this PR: `.fast-boot.json` file at the root of the CLI installation folder.
* After this PR: `~/.balena/cli-module-cache.json` (user's homedir) (`_balena` instead of `.balena` on Windows, see [FAQ](https://github.com/balena-io/balena-cli/blob/master/TROUBLESHOOTING.md#how-do-i-make-the-balena-cli-persist-data-in-another-directory)).

Consider not just `/usr[/local]/lib/balena-cli` (used by the caxa-based installer in the works and the GUI installer for macOS), but also `pkg`'s internal read-only snapshot filesystem and alternative user-defined installation folders like `/opt/balena-cli`.

Basically `fast-boot` never worked properly in such installations or with `pkg`, actually adding a small performance hit instead of a small performance benefit, and none ever noticed (or cared to fix?). This PR is doing something about it.

(An alternative "doing something about it" is to simply remove `fast-boot2` as "not worth the trouble", but on balance and considering it's already being used, I thought this PR was preferable.)

On Linux when `sudo` is used, depending on system configuration, the homedir may be `/root/` instead of the regular user's homedir. It means there may be 2 separate module caches files (root and non-root), but that's OK. Trying to unify them would add complexity that I suggest is not worth the benefit.